Support structs that have a private inner type

This currently runs afoul of private-in-public rules.

```rust
#[macro_use]
extern crate ref_cast;

struct T;

#[derive(RefCast)]
pub struct U(T);
```

```
error[E0446]: private type `T` in public interface
 --> src/main.rs:6:10
  |
6 | #[derive(RefCast)]
  |          ^^^^^^^ can't leak private type
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -18,7 +18,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        rust: [nightly, beta, stable, 1.56.0]
+        rust: [nightly, beta, stable, 1.61.0]
     steps:
       - uses: actions/checkout@v3
       - uses: dtolnay/rust-toolchain@master

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -22,3 +22,6 @@ members = ["derive"]
 
 [package.metadata.docs.rs]
 targets = ["x86_64-unknown-linux-gnu"]
+
+[patch.crates-io]
+ref-cast = { path = "." }

--- a/derive/src/lib.rs
+++ b/derive/src/lib.rs
@@ -3,12 +3,13 @@
 extern crate proc_macro;
 
 use proc_macro::TokenStream;
-use proc_macro2::{Span, TokenStream as TokenStream2, TokenTree};
-use quote::quote;
-use syn::parse::{Nothing, ParseStream};
+use proc_macro2::{Ident, Span, TokenStream as TokenStream2, TokenTree};
+use quote::{quote, quote_spanned};
+use syn::parse::{Nothing, ParseStream, Parser};
 use syn::punctuated::Punctuated;
 use syn::{
-    parse_macro_input, token, Data, DeriveInput, Error, Expr, Field, Path, Result, Token, Type,
+    parenthesized, parse_macro_input, token, Abi, Attribute, Data, DeriveInput, Error, Expr, Field,
+    Generics, Path, Result, Token, Type, Visibility,
 };
 
 /// Derive the `RefCast` trait.
@@ -53,12 +54,160 @@ use syn::{
 #[proc_macro_derive(RefCast, attributes(trivial))]
 pub fn derive_ref_cast(input: TokenStream) -> TokenStream {
     let input = parse_macro_input!(input as DeriveInput);
-    expand(input)
+    expand_ref_cast(input)
         .unwrap_or_else(Error::into_compile_error)
         .into()
 }
 
-fn expand(input: DeriveInput) -> Result<TokenStream2> {
+/// Derive that makes the `ref_cast_custom` attribute able to generate
+/// freestanding reference casting functions for a type.
+///
+/// Please refer to the documentation of
+/// [`#[ref_cast_custom]`][macro@ref_cast_custom] where these two macros are
+/// documented together.
+#[proc_macro_derive(RefCastCustom, attributes(trivial))]
+pub fn derive_ref_cast_custom(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    expand_ref_cast_custom(input)
+        .unwrap_or_else(Error::into_compile_error)
+        .into()
+}
+
+/// Create a function for a RefCast-style reference cast. Call site gets control
+/// of the visibility, function name, argument name, `const`ness, unsafety, and
+/// documentation.
+///
+/// The `derive(RefCast)` macro produces a trait impl, which means the function
+/// names are predefined, and public if your type is public, and not callable in
+/// `const` (at least today on stable Rust). As an alternative to that,
+/// `derive(RefCastCustom)` exposes greater flexibility so that instead of a
+/// trait impl, the casting functions can be made associated functions or free
+/// functions, can be named what you want, documented, `const` or `unsafe` if
+/// you want, and have your exact choice of visibility.
+///
+/// ```rust
+/// use ref_cast::{ref_cast_custom, RefCastCustom};
+///
+/// #[derive(RefCastCustom)]  // does not generate any public API by itself
+/// #[repr(transparent)]
+/// pub struct Frame([u8]);
+///
+/// impl Frame {
+///     #[ref_cast_custom]  // requires derive(RefCastCustom) on the return type
+///     pub(crate) const fn new(bytes: &[u8]) -> &Self;
+///
+///     #[ref_cast_custom]
+///     pub(crate) fn new_mut(bytes: &mut [u8]) -> &mut Self;
+/// }
+///
+/// // example use of the const fn
+/// const FRAME: &Frame = Frame::new(b"...");
+/// ```
+///
+/// The above shows associated functions, but you might alternatively want to
+/// generate free functions:
+///
+/// ```rust
+/// # use ref_cast::{ref_cast_custom, RefCastCustom};
+/// #
+/// # #[derive(RefCastCustom)]
+/// # #[repr(transparent)]
+/// # pub struct Frame([u8]);
+/// #
+/// impl Frame {
+///     pub fn new<T: AsRef<[u8]>>(bytes: &T) -> &Self {
+///         #[ref_cast_custom]
+///         fn ref_cast(bytes: &[u8]) -> &Frame;
+///
+///         ref_cast(bytes.as_ref())
+///     }
+/// }
+/// ```
+#[proc_macro_attribute]
+pub fn ref_cast_custom(args: TokenStream, input: TokenStream) -> TokenStream {
+    let input = TokenStream2::from(input);
+    let expanded = match (|input: ParseStream| {
+        let attrs = input.call(Attribute::parse_outer)?;
+        let vis: Visibility = input.parse()?;
+        let constness: Option<Token![const]> = input.parse()?;
+        let asyncness: Option<Token![async]> = input.parse()?;
+        let unsafety: Option<Token![unsafe]> = input.parse()?;
+        let abi: Option<Abi> = input.parse()?;
+        let fn_token: Token![fn] = input.parse()?;
+        let ident: Ident = input.parse()?;
+        let mut generics: Generics = input.parse()?;
+
+        let content;
+        let paren_token = parenthesized!(content in input);
+        let arg: Ident = content.parse()?;
+        let colon_token: Token![:] = content.parse()?;
+        let from_type: Type = content.parse()?;
+        let _trailing_comma: Option<Token![,]> = content.parse()?;
+        if !content.is_empty() {
+            let rest: TokenStream2 = content.parse()?;
+            return Err(Error::new_spanned(
+                rest,
+                "ref_cast_custom function is required to have a single argument",
+            ));
+        }
+
+        let arrow_token: Token![->] = input.parse()?;
+        let to_type: Type = input.parse()?;
+        generics.where_clause = input.parse()?;
+        let semi_token: Token![;] = input.parse()?;
+
+        let _: Nothing = syn::parse::<Nothing>(args)?;
+
+        Ok(Function {
+            attrs,
+            vis,
+            constness,
+            asyncness,
+            unsafety,
+            abi,
+            fn_token,
+            ident,
+            generics,
+            paren_token,
+            arg,
+            colon_token,
+            from_type,
+            arrow_token,
+            to_type,
+            semi_token,
+        })
+    })
+    .parse2(input.clone())
+    {
+        Ok(function) => expand_function_body(function),
+        Err(parse_error) => {
+            let compile_error = parse_error.to_compile_error();
+            quote!(#compile_error #input)
+        }
+    };
+    TokenStream::from(expanded)
+}
+
+struct Function {
+    attrs: Vec<Attribute>,
+    vis: Visibility,
+    constness: Option<Token![const]>,
+    asyncness: Option<Token![async]>,
+    unsafety: Option<Token![unsafe]>,
+    abi: Option<Abi>,
+    fn_token: Token![fn],
+    ident: Ident,
+    generics: Generics,
+    paren_token: token::Paren,
+    arg: Ident,
+    colon_token: Token![:],
+    from_type: Type,
+    arrow_token: Token![->],
+    to_type: Type,
+    semi_token: Token![;],
+}
+
+fn expand_ref_cast(input: DeriveInput) -> Result<TokenStream2> {
     check_repr(&input)?;
 
     let name = &input.ident;
@@ -125,6 +274,75 @@ fn expand(input: DeriveInput) -> Result<TokenStream2> {
             }
         }
     })
+}
+
+fn expand_ref_cast_custom(input: DeriveInput) -> Result<TokenStream2> {
+    check_repr(&input)?;
+
+    let name = &input.ident;
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
+    let fields = fields(&input)?;
+    let from = only_field_ty(fields)?;
+    let trivial = trivial_fields(fields)?;
+
+    let assert_trivial_fields = if !trivial.is_empty() {
+        Some(quote! {
+            fn __static_assert() {
+                if false {
+                    #(
+                        ::ref_cast::__private::assert_trivial::<#trivial>();
+                    )*
+                }
+            }
+        })
+    } else {
+        None
+    };
+
+    Ok(quote! {
+        unsafe impl #impl_generics ::ref_cast::__private::RefCastCustom<#from> for #name #ty_generics #where_clause {
+            #assert_trivial_fields
+        }
+    })
+}
+
+fn expand_function_body(function: Function) -> TokenStream2 {
+    let Function {
+        attrs,
+        vis,
+        constness,
+        asyncness,
+        unsafety,
+        abi,
+        fn_token,
+        ident,
+        generics,
+        paren_token,
+        arg,
+        colon_token,
+        from_type,
+        arrow_token,
+        to_type,
+        semi_token,
+    } = function;
+
+    let args = quote_spanned! {paren_token.span=>
+        (#arg #colon_token #from_type)
+    };
+
+    quote_spanned! {semi_token.span=>
+        #(#attrs)*
+        #vis #constness #asyncness #unsafety #abi
+        #fn_token #ident #generics #args #arrow_token #to_type {
+            let _ = || {
+                ::ref_cast::__private::ref_cast_custom::<#from_type, #to_type>(#arg);
+            };
+            unsafe {
+                ::ref_cast::__private::transmute::<#from_type, #to_type>(#arg)
+            }
+        }
+    }
 }
 
 fn check_repr(input: &DeriveInput) -> Result<()> {

--- a/src/custom.rs
+++ b/src/custom.rs
@@ -1,0 +1,27 @@
+// Not public API. Use #[derive(RefCastCustom)] and #[ref_cast_custom].
+#[doc(hidden)]
+pub unsafe trait RefCastCustom<From: ?Sized> {
+    fn __static_assert() {}
+}
+
+pub unsafe trait RefCastOkay<From> {}
+
+unsafe impl<'a, From, To> RefCastOkay<&'a From> for &'a To
+where
+    From: ?Sized,
+    To: ?Sized + RefCastCustom<From>,
+{
+}
+
+unsafe impl<'a, From, To> RefCastOkay<&'a mut From> for &'a mut To
+where
+    From: ?Sized,
+    To: ?Sized + RefCastCustom<From>,
+{
+}
+
+pub fn ref_cast_custom<From, To>(_arg: From)
+where
+    To: RefCastOkay<From>,
+{
+}

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -140,13 +140,16 @@
 #![allow(
     clippy::manual_assert,
     clippy::missing_panics_doc,
-    clippy::module_name_repetitions
+    clippy::missing_safety_doc,
+    clippy::module_name_repetitions,
+    clippy::needless_pass_by_value
 )]
 
+mod custom;
 mod layout;
 mod trivial;
 
-pub use ref_cast_impl::RefCast;
+pub use ref_cast_impl::{ref_cast_custom, RefCast, RefCastCustom};
 
 /// Safely cast `&T` to `&U` where the struct `U` contains a single field of
 /// type `T`.
@@ -177,6 +180,8 @@ pub trait RefCast {
 // Not public API.
 #[doc(hidden)]
 pub mod __private {
+    pub use crate::custom::{ref_cast_custom, RefCastCustom};
     pub use crate::layout::{assert_layout, Layout, LayoutUnsized};
     pub use crate::trivial::assert_trivial;
+    pub use core::mem::transmute;
 }

--- a/tests/ui/extra-arg.rs
+++ b/tests/ui/extra-arg.rs
@@ -1,0 +1,12 @@
+use ref_cast::{ref_cast_custom, RefCastCustom};
+
+#[derive(RefCastCustom)]
+#[repr(transparent)]
+pub struct Thing(String);
+
+impl Thing {
+    #[ref_cast_custom]
+    pub fn ref_cast(s: &String, wat: i32) -> &Self;
+}
+
+fn main() {}

--- a/tests/ui/extra-arg.stderr
+++ b/tests/ui/extra-arg.stderr
@@ -1,0 +1,13 @@
+error: ref_cast_custom function is required to have a single argument
+ --> tests/ui/extra-arg.rs:9:33
+  |
+9 |     pub fn ref_cast(s: &String, wat: i32) -> &Self;
+  |                                 ^^^^^^^^
+
+error: associated function in `impl` without body
+ --> tests/ui/extra-arg.rs:9:5
+  |
+9 |     pub fn ref_cast(s: &String, wat: i32) -> &Self;
+  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-
+  |                                                   |
+  |                                                   help: provide a definition for the function: `{ <body> }`

--- a/tests/ui/function-body.rs
+++ b/tests/ui/function-body.rs
@@ -1,0 +1,12 @@
+use ref_cast::{ref_cast_custom, RefCastCustom};
+
+#[derive(RefCastCustom)]
+#[repr(transparent)]
+pub struct Thing(String);
+
+impl Thing {
+    #[ref_cast_custom]
+    pub fn ref_cast(s: &String) -> &Self {}
+}
+
+fn main() {}

--- a/tests/ui/function-body.stderr
+++ b/tests/ui/function-body.stderr
@@ -1,0 +1,13 @@
+error: expected `;`
+ --> tests/ui/function-body.rs:9:42
+  |
+9 |     pub fn ref_cast(s: &String) -> &Self {}
+  |                                          ^
+
+error[E0308]: mismatched types
+ --> tests/ui/function-body.rs:9:36
+  |
+9 |     pub fn ref_cast(s: &String) -> &Self {}
+  |            --------                ^^^^^ expected `&Thing`, found `()`
+  |            |
+  |            implicitly returns `()` as its body has no tail or `return` expression

--- a/tests/ui/impl-trait.rs
+++ b/tests/ui/impl-trait.rs
@@ -1,0 +1,15 @@
+use ref_cast::{ref_cast_custom, RefCastCustom};
+
+#[derive(RefCastCustom)]
+#[repr(transparent)]
+pub struct Thing(str);
+
+impl Thing {
+    #[ref_cast_custom]
+    pub fn ref_cast(s: impl AsRef<str>) -> &Self;
+
+    #[ref_cast_custom]
+    pub fn ref_cast2(s: &impl AsRef<str>) -> &Self;
+}
+
+fn main() {}

--- a/tests/ui/impl-trait.stderr
+++ b/tests/ui/impl-trait.stderr
@@ -1,0 +1,23 @@
+error[E0106]: missing lifetime specifier
+ --> tests/ui/impl-trait.rs:9:44
+  |
+9 |     pub fn ref_cast(s: impl AsRef<str>) -> &Self;
+  |                                            ^ expected named lifetime parameter
+  |
+  = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
+help: consider using the `'static` lifetime
+  |
+9 |     pub fn ref_cast(s: impl AsRef<str>) -> &'static Self;
+  |                                             +++++++
+
+error[E0562]: `impl Trait` only allowed in function and inherent method return types, not in path
+ --> tests/ui/impl-trait.rs:9:24
+  |
+9 |     pub fn ref_cast(s: impl AsRef<str>) -> &Self;
+  |                        ^^^^^^^^^^^^^^^
+
+error[E0562]: `impl Trait` only allowed in function and inherent method return types, not in path
+  --> tests/ui/impl-trait.rs:12:26
+   |
+12 |     pub fn ref_cast2(s: &impl AsRef<str>) -> &Self;
+   |                          ^^^^^^^^^^^^^^^

--- a/tests/ui/no-custom.rs
+++ b/tests/ui/no-custom.rs
@@ -1,0 +1,11 @@
+use ref_cast::ref_cast_custom;
+
+#[repr(transparent)]
+pub struct Thing(String);
+
+impl Thing {
+    #[ref_cast_custom]
+    pub fn ref_cast(s: &String) -> &Self;
+}
+
+fn main() {}

--- a/tests/ui/no-custom.stderr
+++ b/tests/ui/no-custom.stderr
@@ -1,0 +1,15 @@
+error[E0277]: the trait bound `Thing: RefCastCustom<String>` is not satisfied
+ --> tests/ui/no-custom.rs:8:36
+  |
+8 |     pub fn ref_cast(s: &String) -> &Self;
+  |                                    ^^^^^ the trait `RefCastCustom<String>` is not implemented for `Thing`
+  |
+  = help: the following other types implement trait `ref_cast::custom::RefCastOkay<From>`:
+            <&'a To as ref_cast::custom::RefCastOkay<&'a From>>
+            <&'a mut To as ref_cast::custom::RefCastOkay<&'a mut From>>
+  = note: required for `&Thing` to implement `ref_cast::custom::RefCastOkay<&String>`
+note: required by a bound in `ref_cast_custom`
+ --> src/custom.rs
+  |
+  |     To: RefCastOkay<From>,
+  |         ^^^^^^^^^^^^^^^^^ required by this bound in `ref_cast_custom`

--- a/tests/ui/private.rs
+++ b/tests/ui/private.rs
@@ -1,0 +1,19 @@
+use ref_cast::{ref_cast_custom, RefCast, RefCastCustom};
+
+#[derive(RefCast, RefCastCustom)]
+#[repr(transparent)]
+pub struct Public {
+    private: Private,
+}
+
+struct Private;
+
+impl Public {
+    #[ref_cast_custom]
+    fn ref_cast(private: &Private) -> &Public;
+
+    #[ref_cast_custom]
+    fn ref_cast_mut(private: &mut Private) -> &mut Public;
+}
+
+fn main() {}

--- a/tests/ui/private.stderr
+++ b/tests/ui/private.stderr
@@ -1,0 +1,10 @@
+error[E0446]: private type `Private` in public interface
+ --> tests/ui/private.rs:3:10
+  |
+3 | #[derive(RefCast, RefCastCustom)]
+  |          ^^^^^^^ can't leak private type
+...
+9 | struct Private;
+  | -------------- `Private` declared as private
+  |
+  = note: this error originates in the derive macro `RefCast` (in Nightly builds, run with -Z macro-backtrace for more info)

--- a/tests/ui/short-lifetime.rs
+++ b/tests/ui/short-lifetime.rs
@@ -1,0 +1,12 @@
+use ref_cast::{ref_cast_custom, RefCastCustom};
+
+#[derive(RefCastCustom)]
+#[repr(transparent)]
+pub struct Thing(String);
+
+impl Thing {
+    #[ref_cast_custom]
+    pub fn ref_cast<'a>(s: &String) -> &'a Self;
+}
+
+fn main() {}

--- a/tests/ui/short-lifetime.stderr
+++ b/tests/ui/short-lifetime.stderr
@@ -1,0 +1,7 @@
+error[E0621]: explicit lifetime required in the type of `s`
+ --> tests/ui/short-lifetime.rs:9:48
+  |
+9 |     pub fn ref_cast<'a>(s: &String) -> &'a Self;
+  |                            -------             ^ lifetime `'a` required
+  |                            |
+  |                            help: add explicit lifetime `'a` to the type of `s`: `&'a String`

